Map vector types to arrays and pass them by value

gnzlbg opened the pull request:
This PR maps vector types to arrays and passes them by value (that is, they are passed to C as `[T; N]`.  This already allows defining them as a blacklisted opaque type, such that the user can provide its own definition from e.g. `std::arch`. 

A subsequent PR should map vector types to unit structs with a private member:

```rust
#[repr(align(16))] pub struct __m128([f32; 4]); 
```

to make their alignment at least be correct. This should allow transmuting `std::arch` types to these types properly. 

Once that is done, we probably want to detect the canonical vector types (e.g. `__m128`) and pull in the type from `std`/`core`::arch instead.

gnzlbg commented:
The builds failures appear spurious. 

I'm using this branch to finally build `sleef-sys` successfully (https://github.com/gnzlbg/sleef-sys), so I would prefer to focus into getting some sort of this merged in an usable way that breaks nobody first, and ironing out other issues later.

emilio commented on src/ir/ty.rs:
```diff
@@ -627,6 +630,9 @@ pub enum TypeKind {
     TemplateAlias(TypeId, Vec<TypeId>),
 
+    /// A packed vector type: element type, number of elements, vector alignment.
+    Vector(TypeId, usize, usize),
```
Can we name the fields if we need to keep the alignment?

emilio commented on src/ir/ty.rs:
```diff
@@ -1148,12 +1154,15 @@ impl Type {
+                        None,
+                        ctx,
+                    ).expect("Not able to resolve vector element?");
+                    TypeKind::Vector(inner, ty.num_elements().unwrap(), ty.align())
```
Is `ty.align()` the same as `layout().align`? If so, there's no need to store it in the vector.

emilio commented on src/ir/analysis/derive_partialeq_or_partialord.rs:
```diff
@@ -199,6 +199,14 @@ impl<'ctx> CannotDerivePartialEqOrPartialOrd<'ctx> {
+            TypeKind::Vector(..) => {
+                // FIXME: vectors always can derive PartialEq, but they can
+                // never derive PartialOrd. 
+                trace!(
```
nit: This fits in one line.

emilio commented on src/ir/analysis/derive_hash.rs:
```diff
@@ -202,6 +202,18 @@ impl<'ctx> MonotoneFramework for CannotDeriveHash<'ctx> {
+                    );
+                    return self.insert(id);
+                }
+                assert!(len != 0);
```
nit: `assert_ne!` gives a nicer error message.

emilio commented on src/codegen/mod.rs:
```diff
@@ -3055,6 +3056,12 @@ impl TryToRustTy for Type {
                     [ #ty ; #len ]
                 })
             }
+            TypeKind::Vector(item, len, _) => {
```
We can merge this path with the one above for now if you want.

emilio commented on src/ir/analysis/derive_partialeq_or_partialord.rs:
```diff
@@ -199,6 +199,14 @@ impl<'ctx> CannotDerivePartialEqOrPartialOrd<'ctx> {
             }
+            TypeKind::Vector(..) => {
+                // FIXME: vectors always can derive PartialEq, but they can
+                // never derive PartialOrd. 
```
Maybe point to the link you sent over IRC for reference about why?

gnzlbg commented:
@emilio I think I have addressed all the comments. I've removed the alignment for now (it is unused as you said), and fixed the other changes suggested.

emilio commented:
@bors-servo r+

Sure, will publish as soon as it merges :)

bors-servo commented:
:pushpin: Commit 6e7d9be has been approved by `emilio`

<!-- @bors-servo r=emilio 6e7d9bebd8c95a8ca7088cf02c17aea811d20783 -->

bors-servo commented:
:hourglass: Testing commit 6e7d9bebd8c95a8ca7088cf02c17aea811d20783 with merge 40d24bb085a1a0bd11f7cf4dd87c52de529302ca...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang-nursery/rust-bindgen/builds/416147362?utm_source=github_status&utm_medium=notification)
Approved by: emilio
Pushing 40d24bb085a1a0bd11f7cf4dd87c52de529302ca to master...

emilio commented:
0.38.0 should be up now, sorry for the lag!

gnzlbg commented:
Thanks!
